handlers/secrets.go

```diff
@@ -28,30 +29,67 @@ const (
 
 // MakeSecretHandler makes a handler for Create/List/Delete/Update of
 //secrets in the Kubernetes API
-func MakeSecretHandler(namespace string, kube kubernetes.Interface) http.HandlerFunc {
+func MakeSecretHandler(defaultNamespace string, kube kubernetes.Interface) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		if r.Body != nil {
 			defer r.Body.Close()
 		}
 
-		client := kube.CoreV1().Secrets(namespace)
+		lookupNamespace, err, errCode := getLookupNamespace(defaultNamespace, r)
+		if err != nil {
+			http.Error(w, err.Error(), errCode)
+		}
+
+		client := kube.CoreV1().Secrets(lookupNamespace)
 
 		switch r.Method {
 		case http.MethodGet:
 			getSecretsHandler(client, w, r)
 		case http.MethodPost:
-			createSecretHandler(client, namespace, w, r)
+			createSecretHandler(client, lookupNamespace, w, r)
 		case http.MethodPut:
-			replaceSecretHandler(client, namespace, w, r)
+			replaceSecretHandler(client, lookupNamespace, w, r)
 		case http.MethodDelete:
-			deleteSecretHandler(client, namespace, w, r)
+			deleteSecretHandler(client, lookupNamespace, w, r)
 		default:
 			w.WriteHeader(http.StatusBadRequest)
 			return
 		}
 	}
 }
 
+// getLookupNamespace returns the namespace specified in any request type
+func getLookupNamespace(defaultNamespace string, r *http.Request) (string, error, int) {
```

Comment:
In general, I would prefer the handler to decide which error code to return instead.  One, I generally expect `error` to be the last item returned from a method

```go
func getLookupNamespace(defaultNamespace string, r *http.Request) (string, error) { }
```

And second, I think the separation of concerns is cleaner when a handler catches errors and then decides what the appropriate code is. 

This method could return a sentinel error (see https://dave.cheney.net/2019/06/10/constant-time for an example if you are not familiar) for each error case and then have the handler can easily use a switch to decide the correct status code.  This would then be cleaner for us to reuse the code in the future in cases that are not handlers, for example, in the openfaas-operator

Reply:
Thanks for the advice and the article, really good read :)

I refactored the code so the handler catches the error, based on the error string (_is there a better way than checking the string_?).